### PR TITLE
Use try-with-resource and implement Closeable in MutableProjectRegistry

### PR DESCRIPTION
**Pull request by HannesWell**
This PR aims to reduce the number of close operations in finally-blocks of try-constructs.
For this I use try-with-resources where applicable.
While in some cases this is only a syntactical difference, in some places this also has the effect that Exceptions that are thrown while closing the resource are not ignored any more. But I think this is OK, because first I think Exceptions while closing is rare and second I think that those Exceptions should be handled equivalently to for example cases where an exception is thrown while opening the resource.

At some locations I also simplified the code within the try block using Java's Files class.
Furthermore I let `MutableProjectRegistry` implement the AutoCloseable interface, because it already had a `close()` method that was only called in the finally-block of try statements. So I think this is suitable.

I started the Commit a while ago but didn't finish it yet. Alexanders recent PR(s) that also leverage try-with-resources remembered me to complete it, in order to avoid duplicated work. The currently pending PR of him also contains some try-with-resources applications that likely overlap with this PR, but the result should be the same.

### CODE REVIEW

**Comment by akurtakov** [comment]
There seem to be some overlap with my yesterday fixes. Would you please fix the conflicts?

**Comment by HannesWell** [comment]
I have rebased the commit, resolved the conflicts and forced pushed it.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/ProjectRegistryReader.java [hunk 1]
```diff
@@ -65,53 +63,54 @@ public ProjectRegistryReader(File stateLocationDir) {
   public ProjectRegistry readWorkspaceState(final ProjectRegistryManager managerImpl) {
     if(stateFile.exists()) {
       final PackageAdmin packageAdmin = getPackageAdmin();
-      ObjectInputStream is = null;
-      try {
-        is = new ObjectInputStream(new BufferedInputStream(new FileInputStream(stateFile))) {
```
Comment (HannesWell) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/ProjectRegistryReader.java:70]: The creation of the ObjectInputStream was simply extracted into an own class to improve readability.
The same was done for the ObjectOutputStream below.
I think an anonymous class within a try-with-resources would not be very readable.

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/actions/SelectionUtil.java [hunk 2]
```diff
@@ -266,19 +263,13 @@ public static MavenProject getMavenProject(IEditorInput editorInput, IProgressMo
...
         File tempPomFile = null;
-        InputStream is = null;
-        OutputStream os = null;
-        try {
+        try (InputStream is = storage.getContents()) {
           tempPomFile = File.createTempFile("maven-pom", ".pom"); //$NON-NLS-1$ //$NON-NLS-2$
-          os = new FileOutputStream(tempPomFile);
-          is = storage.getContents();
-          IOUtil.copy(is, os);
+          Files.copy(storage.getContents(), tempPomFile.toPath());
```
Comment (mickaelistria) [inline org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/actions/SelectionUtil.java:268]: should this be
```suggestion
          Files.copy(is, tempPomFile.toPath());
```

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MutableProjectRegistry.java [hunk 3]
```diff
@@ -35,7 +35,7 @@
  *
  * @author igor
  */
-public class MutableProjectRegistry extends BasicProjectRegistry implements IProjectRegistry {
+public class MutableProjectRegistry extends BasicProjectRegistry implements IProjectRegistry, AutoCloseable {
```
Comment (mickaelistria) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/MutableProjectRegistry.java:38]: Good idea.

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/ProjectRegistryReader.java (hunk 1 above)
Comment (mickaelistria) [reply org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/project/registry/ProjectRegistryReader.java:70]: :+1:

File: org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/actions/SelectionUtil.java (hunk 2 above)
Comment (HannesWell) [reply org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/actions/SelectionUtil.java]: Yes you are right. Thanks for pointing out!

**Pull request merged**